Prevent units in tunnels/underground from attacking

reaperrr opened the pull request:
Fixes #12917.

reaperrr commented:
Together with #12921, this is probably all I can contribute towards solving the TS public alpha blockers (except reviewing PRs adressing the other blockers, of course).

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -382,6 +382,8 @@
 		ReferencePoint: Bottom, Right
 		RequiresCondition: hospitalheal
 	RevealOnFire:
+	RejectsOrders:
```
This is a bad idea for now - if units stop inside tunnels (because Move sucks) then it won't be possible to recover them.

reaperrr replied:
How about adding `Except: Move`? That way at least Move orders are still possible, while anything potentially problematic - especially `Stop` - is not.

pchote replied:
Can we start by only rejecting Stop?

reaperrr replied:
I don't mind, it's just that I'd have to list every other order here, unless we add an explicit `Reject` field to this trait or an explicit `AcceptsOrders` counter-part trait.

pchote replied:
> add an explicit `Reject` field to this trait

Sounds good to me.

GraionDilach replied:
Infact, I'd prefer if we could allow multiple `RejectOrders` traits on actors along with that `Reject` field, would be helpful for modders.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs:
```diff
@@ -93,6 +93,10 @@ public override Activity Tick(Actor self)
 
+				var mobile = move as Mobile;
+				if (mobile != null && !mobile.CanInteractWithGroundLayer(self))
+					return this;
```
Are you sure about this?  It looks like this will make the unit stop, instead of repathing.

reaperrr replied:
No, I'm not sure about this. I'm not sure how to make the unit repath instead either, though.

reaperrr commented:
Updated.

abcdefg30 commented on OpenRA.Mods.Common/Traits/RejectsOrders.cs:
```diff
@@ -35,8 +41,8 @@ public static class RejectsOrdersExts
-			var r = self.TraitOrDefault<RejectsOrders>();
-			return r == null || r.IsTraitDisabled || r.Except.Contains(orderString);
+			var r = self.TraitsImplementing<RejectsOrders>().Where(Exts.IsTraitEnabled).ToList();
+			return !r.Any() || (r.Any(t => t.Reject.Count > 0 && !t.Reject.Contains(orderString))) || (r.Any(t => t.Except.Contains(orderString)));
```
> OpenRA.Mods.Common/Traits/RejectsOrders.cs:L45: [SA1119] The line contains unnecessary parenthesis.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs:
```diff
@@ -93,6 +93,11 @@ public override Activity Tick(Actor self)
 				if (self.IsDisabled())
 					return this;
 
+				// If target moves inside tunnel, don't follow
```
Do we really want this behaviour?  I imagine that ideally we would follow them through the tunnel.

pchote commented on OpenRA.Mods.Common/Traits/RejectsOrders.cs:
```diff
@@ -17,14 +17,19 @@ namespace OpenRA.Mods.Common.Traits
 	{
-		[Desc("Possible values include Attack, AttackMove, Guard, Move.")]
+		[Desc("Explicit list of rejected orders. Leave empty to reject all minus those listed under Except. ",
+			"Possible values include Attack, AttackMove, Guard, Move, Stop.")]
```
This list is just asking for bitrot, IMO.

GraionDilach replied:
It already bitrot with things like the deploy commands.

reaperrr replied:
Well, 'include' implies that this list is not complete, but I guess we might as well remove it.

pchote replied:
Might be worth adding a mechanism to include order names as part of the autogenerated trait docs.

pchote commented on OpenRA.Mods.Common/Traits/RejectsOrders.cs:
```diff
@@ -35,8 +41,8 @@ public static class RejectsOrdersExts
-			var r = self.TraitOrDefault<RejectsOrders>();
-			return r == null || r.IsTraitDisabled || r.Except.Contains(orderString);
+			var r = self.TraitsImplementing<RejectsOrders>().Where(Exts.IsTraitEnabled).ToList();
+			return !r.Any() || r.Any(t => t.Reject.Count > 0 && !t.Reject.Contains(orderString)) || r.Any(t => t.Except.Contains(orderString));
```
Style nit: Prefer `.Any()` over `.Count > 0`.

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -566,6 +569,9 @@
 		RequiresCondition: !inside-tunnel
 	RevealOnFire:
+	RejectsOrders@tunnel:
+		Reject: Stop
```
IMO we would ideally not reject, but defer these orders so that they stop/deploy/etc as soon as they exit the tunnel.  This should be possible, in principle, by making the `Move` activity uncancellable if `!mobile.CanInteractWithGroundLayer`.  Knowing the current state of the `Move` code, I would be surprised if it was really that simple (ping: @obrakmann).

`RejectsOrders` is a good enough stopgap for now, and the improvements to that trait are certainly worthwhile for other use cases.

pchote replied:
FYI there are a few other ways to stop a unit in a tunnel, including but probably not limited to:

* Unloading cargo
* Deploy an MCV/tick tank/etc
* Issuing a move order that fails (e.g. into water)

All of these would go away if the activity cancelling idea works.  If not, we can cover the first two cases by rejecting more orders, and all the other cases by having an idle activity that automatically moves idle actors to the nearest tunnel entrance.

reaperrr replied:
> Issuing a move order that fails (e.g. into water)

Wouldn't it make more sense to generally ignore invalid move orders completely and simply continue with the last valid order?

GraionDilach replied:
Couldn'tv we rejectv all orders within tunnels for now? Basically that's what the WW games did.

pchote replied:
Probably, but that's a whole other set of headaches to the main problem here.

GraionDilach replied:
I won't argue on that longterm, but I believe this PR should simply aim for the WW restoration.

pchote replied:
That is probably the most efficient way forward.  The "proper" fix that I have in mind will depend on the keepQueue flag from #12919, so it would make more sense to get this and that merged and then revisit the idea again in its own PR.

pchote replied:
Actually, I take that back.  The following should be enough, but will of course need some proper testing:
```diff
diff --git a/OpenRA.Mods.Common/Activities/Move/Move.cs b/OpenRA.Mods.Common/Activities/Move/Move.cs
index e7418d6e8..2ce6e5cda 100644
--- a/OpenRA.Mods.Common/Activities/Move/Move.cs
+++ b/OpenRA.Mods.Common/Activities/Move/Move.cs
@@ -142,7 +142,7 @@ List<CPos> EvalPath()
 
                public override Activity Tick(Actor self)
                {
-                       if (IsCanceled)
+                       if (IsCanceled && mobile.CanInteractWithGroundLayer(self))
                                return NextActivity;
 
                        if (mobile.IsTraitDisabled)
@@ -277,12 +277,6 @@ void SanityCheckPath(Mobile mobile)
                        return Pair.New(nextCell, subCell);
                }
 
-               public override bool Cancel(Actor self)
-               {
-                       path = NoPath;
-                       return base.Cancel(self);
-               }
-
                public override IEnumerable<Target> GetTargets(Actor self)
                {
                        if (path != null)
```

obrakmann replied:
Removing `Move`'s `Cancel` override will probably break it, I'm guessing, due to the `path = NoPath` line.

Using the `keepQueue` flag with `Move` won't work out-of-the-box either, since `Move` doesn't use inner or child activities yet, but queues the other `MovePart`s as next activities. So if you cancel `Move` with the `keepQueue` flag at just the wrong time, it would end up running the `MovePart` anyway.

I have a WIP branch that re-writes `Move` using child activities, but it isn't done yet.

pchote replied:
As far as I can tell setting `NoPath` doesn't have any effect on the behaviour of the trait.  The trait cancelled check at the start of `Tick` disables all uses of `path` after it would have been assigned.

obrakmann replied:
`Move.Cancel` is also called from the `MovePart`s' `Cancel` methods, and they're using the `path` via their `parent` argument.

pchote replied:
Right, well that makes this only slightly less trivial: https://github.com/OpenRA/OpenRA/commit/2e8f8f8d08daf698137666095bae1af5e066e322?w=1

pchote reviewed (changes requested):
Looks good overall.  See below for couple of minor requests, one major one that might not work, and some rambling:

reaperrr committed:
Prevent attacking while in tunnel/underground

reaperrr commented:
Rebased and updated.

I kept the improvements to `RejectsOrders`, but the actual fix is now something much simpler, `Attack*.CanAttack` now simply returns `false` while inside a tunnel layer.

pchote commented:
This also needs to apply for subterranian units.

reaperrr commented:
Updated, also fixed the style nits I previously forgot to fix.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -80,6 +80,10 @@ protected virtual bool CanAttack(Actor self, Target target)
 				return false;
 
+			var layer = self.Location.Layer;
+			if (layer == CustomMovementLayerType.Tunnel || layer == CustomMovementLayerType.Subterranean)
```
Please use `mobile.CanInteractWithGroundLayer`, which is designed for cases like this.

pchote replied:
Unfortunately this makes units drop their target as soon as it enters a tunnel.  I don't think the older approach had this problem, so could you try bringing that back?

pchote replied:
Actually, scratch that.  I tested the older version and both that and my earlier AttackFrontal changes do the same thing.  :+1: for the current approach if you can `TraitOrDefault<Mobile>` and check `CanInteractWithGroundLayer`.

reaperrr replied:
There's one other thing I could try to get the desired behavior, namely treating targets on non-interactable layers as if they were 'out of range' in the `Attack` activity. That would be a little hacky, though.

pchote replied:
That's what we do in https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/Attack.cs#L105 (for AttackFrontal) and what I thought your old implementation was doing, but it still cancels the target.

pchote replied:
I just realized that i'm confusing matters here.  I'm saying that for unit A targeting unit B, unit A drops its target when unit B enters a tunnel.  These changes are all about what happens when unit A enters a tunnel, so ^ is a different issue.

reaperrr commented:
Updated.

reaperrr commented:
Updated, split `RejectsOrders` changes to #13202 as I suspect they might be putting off reviewers from approving this PR.

atlimit8 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9a292cbf5f1931605de642f6e2e73e6bf2c412b5...8d2b35c1a931a75a6d85edf6ebaa436049275f43)
